sw/device/lib/crypto/drivers/kmac.c

```diff
@@ -37,84 +319,232 @@ static kmac_error_t wait_status_bit(uint32_t bit_position, uint32_t bit_value) {
   }
 }
 
+/**
+ * Encode a given integer as byte array and return its size along with it.
+ *
+ * This is a common procedure that can be used to implement both `left_encode`
+ * and `right_encode` functions defined in NIST SP 800-185. Given an integer
+ * `value` it returns its encoding as a byte array in `encoding_buf`. Meanwhile,
+ * `encoding_header` keeps the size of `encoding_buf`. Later the two can be
+ * combined as below:
+ *
+ * left_encode(`value`) = `encoding_header` || `encoding_buf`
+ * right_encode(`value`) = `encoding_buf` || `encoding_header`
+ *
+ * The implementation is hardcoded and assumes that `value` is less than 2^32.
+ * Otherwise `kKmacDigestLenTooLongError` is returned.
+ *
+ * @param value Integer to be encoded.
+ * @param encoding_buf The output byte array representing `value`.
+ * @param encoding_header The number of bytes written to `encoded_value`.
+ * @return Error code.
+ */
 OT_WARN_UNUSED_RESULT
-kmac_error_t kmac_init(kmac_operation_t operation) {
+static kmac_error_t little_endian_encode(const size_t value,
+                                         uint8_t *encoding_buf,
+                                         uint8_t *encoding_header) {
+  if (encoding_buf == NULL || encoding_header == NULL) {
+    return kKmacArgsError;
+  }
+
+  *encoding_buf++ = value & 0xff;
+  if (value < 0xff) {
+    *encoding_header = 1;
+  } else if (value < 0xffff) {
+    *encoding_header = 2;
+    *encoding_buf++ = (value >> 8) & 0xff;
+  } else if (value < 0xffffff) {
+    *encoding_header = 3;
+    *encoding_buf++ = (value >> 8) & 0xff;
+    *encoding_buf++ = (value >> 16) & 0xff;
+  } else if (value < 0xffffffff) {
+    *encoding_header = 4;
+    *encoding_buf++ = (value >> 8) & 0xff;
+    *encoding_buf++ = (value >> 16) & 0xff;
+    *encoding_buf++ = (value >> 24) & 0xff;
+  } else {
+    // We probably don't need to consider more than 4 bytes for two reasons:
+    // 1) size_t is 4 bytes
+    // 2) One needs to request digest_len = 2^32 during KMAC operation
+    return kKmacDigestLenTooLongError;
+  }
```

Comment:
I think this would be clearer and more robust as a loop, e.g.
```c
uint8_t len = 0;
while (value > 0) {
  encoding_buf[len] = value & UINT8_MAX
  value >>= 8;
  len++;
}
*encoding_header = len;
```
This way if the size limits change, the code won't break.

Reply:
Replaced it with a do-while block (as `value`=0 also needs `*encoding_header = 1` and `encoding_buf[0] = 0`.
https://github.com/lowRISC/opentitan/pull/16180/files#diff-e3a8a239a1792c5cff351885260cf2dbc860356c60b87e34d1cf5387e65db687R346-R352

I also increased the size of the allocated buffer to 256 bytes in the outer scope (max allowed size according to NIST SP 800-185 `left_encode` or `right_encode` functions.)